File: scripts/sun-times.py
#!/usr/bin/env python3
import math, sys
from datetime import datetime

# Lapeer, MI
LAT = 43.05
LNG = -83.32

def sun_times(date):
    """Calculate sunrise/sunset in local decimal hours using NOAA algorithm."""
    import time
    is_dst = time.daylight and time.localtime().tm_isdst > 0
    tz_offset = -(time.timezone if not is_dst else time.altzone) / 3600.0

    day_of_year = date.timetuple().tm_yday

    # Fractional year (radians)
    gamma = (2 * math.pi / 365) * (day_of_year - 1)

    # Equation of time (minutes)
    eqtime = 229.18 * (0.000075 + 0.001868 * math.cos(gamma)
             - 0.032077 * math.sin(gamma)
             - 0.014615 * math.cos(2 * gamma)
             - 0.040849 * math.sin(2 * gamma))

    # Solar declination (radians)
    decl = (0.006918 - 0.399912 * math.cos(gamma)
            + 0.070257 * math.sin(gamma)
            - 0.006758 * math.cos(2 * gamma)
            + 0.000907 * math.sin(2 * gamma)
            - 0.002697 * math.cos(3 * gamma)
            + 0.00148 * math.sin(3 * gamma))

    lat_rad = math.radians(LAT)

    # Hour angle (degrees)
    cos_ha = (math.cos(math.radians(90.833)) / (math.cos(lat_rad) * math.cos(decl))
              - math.tan(lat_rad) * math.tan(decl))
    cos_ha = max(-1, min(1, cos_ha))
    ha = math.degrees(math.acos(cos_ha))

    # Sunrise and sunset in minutes from midnight UTC
    sunrise_utc = 720 - 4 * (LNG + ha) - eqtime
    sunset_utc = 720 - 4 * (LNG - ha) - eqtime

    # Convert to local hours
    sunrise_local = (sunrise_utc / 60.0) + tz_offset
    sunset_local = (sunset_utc / 60.0) + tz_offset

    return sunrise_local, sunset_local

if __name__ == '__main__':
    today = datetime.now()
    sunrise, sunset = sun_times(today)

    if len(sys.argv) > 1:
        mode = sys.argv[1]
        if mode == 'sunrise':
            h = int(sunrise)
            m = int((sunrise - h) * 60)
            print(f"{h}:{m:02d}")
        elif mode == 'sunset':
            h = int(sunset)
            m = int((sunset - h) * 60)
            print(f"{h}:{m:02d}")
        elif mode == 'check-run':
            hour = today.hour + today.minute / 60.0
            sys.exit(0 if (sunrise - 1) <= hour <= (sunset + 1) else 1)
        elif mode == 'check-coop':
            hour = today.hour + today.minute / 60.0
            if hour >= (sunset - 1) or hour <= (sunrise + 1):
                sys.exit(0)
            else:
                sys.exit(1)
    else:
        sr_h = int(sunrise)
        sr_m = int((sunrise - sr_h) * 60)
        ss_h = int(sunset)
        ss_m = int((sunset - ss_h) * 60)
        print(f"sunrise={sr_h}:{sr_m:02d} sunset={ss_h}:{ss_m:02d}")

File: public/sun-times.js
// Sunrise/sunset calculator for the farm (Lapeer, MI), ported from
// scripts/sun-times.py (NOAA solar calculation, low-accuracy Fourier form:
// https://gml.noaa.gov/grad/solcalc/solareqns.PDF). Everything is UTC epoch
// milliseconds — no timezone or DST math anywhere.
//
// Loads as a plain browser script (defines window.SunTimes) and as a CommonJS
// module for node:test (module.exports tail at the bottom).
(function (root) {
  'use strict';

  var FARM_LAT = 43.05;
  var FARM_LNG = -83.32;

  var MS_PER_MINUTE = 60 * 1000;
  var MS_PER_DAY = 24 * 60 * MS_PER_MINUTE;
  var MONTHS_PER_YEAR = 12;
  var DAYS_PER_YEAR = 365; // NOAA low-accuracy form ignores leap years
  var SOLAR_NOON_UTC_MINUTES = 720; // 12:00 UTC, reference point of the NOAA formula
  var MINUTES_PER_LONGITUDE_DEGREE = 4; // Earth rotates 1 degree every 4 minutes
  // Sun's center 0.833 degrees below horizon at rise/set: 0.567 atmospheric
  // refraction + 0.266 solar semi-diameter (NOAA "official" sunrise/sunset).
  var OFFICIAL_ZENITH_DEGREES = 90.833;
  // 1440 minutes/day divided by 2*pi radians: converts the equation-of-time
  // Fourier series (radians of hour angle) to minutes.
  var EQ_OF_TIME_MINUTES_PER_RADIAN = 229.18;

  function toRadians(degrees) {
    return (degrees * Math.PI) / 180;
  }

  function toDegrees(radians) {
    return (radians * 180) / Math.PI;
  }

  /**
   * Sunrise/sunset for one UTC calendar day at the farm.
   *
   * Month is 1-BASED (1 = January ... 12 = December), unlike JS Date's 0-based
   * months — e.g. getSunWindowUtc(2026, 7, 10) is July 10, 2026 (UTC).
   *
   * Returned values are absolute epoch milliseconds. sunsetMs may land on the
   * NEXT UTC calendar day: in Michigan summer, sunset (~21:10 EDT) is ~01:10
   * UTC the following day. That is correct — callers compare absolute ms.
   *
   * @param {number} utcYear  full year, e.g. 2026
   * @param {number} utcMonth 1-based month (1-12)
   * @param {number} utcDay   day of month (1-31)
   * @returns {{ sunriseMs: number, sunsetMs: number }}
   */
  function getSunWindowUtc(utcYear, utcMonth, utcDay) {
    if (!Number.isFinite(utcYear) || !Number.isFinite(utcMonth) || !Number.isFinite(utcDay)) {
      throw new TypeError('getSunWindowUtc: year/month/day must be finite numbers');
    }
    if (utcMonth < 1 || utcMonth > MONTHS_PER_YEAR) {
      throw new RangeError('getSunWindowUtc: month must be 1-12 (1-based), got ' + utcMonth);
    }

    var dayStartMs = Date.UTC(utcYear, utcMonth - 1, utcDay);
    var yearStartMs = Date.UTC(utcYear, 0, 1);
    var dayOfYear = Math.round((dayStartMs - yearStartMs) / MS_PER_DAY) + 1;

    // Fractional year (radians)
    var gamma = ((2 * Math.PI) / DAYS_PER_YEAR) * (dayOfYear - 1);

    // Equation of time (minutes) — NOAA Fourier coefficients
    var eqTimeMinutes = EQ_OF_TIME_MINUTES_PER_RADIAN * (
      0.000075 +
      0.001868 * Math.cos(gamma) -
      0.032077 * Math.sin(gamma) -
      0.014615 * Math.cos(2 * gamma) -
      0.040849 * Math.sin(2 * gamma)
    );

    // Solar declination (radians) — NOAA Fourier coefficients
    var declinationRad =
      0.006918 -
      0.399912 * Math.cos(gamma) +
      0.070257 * Math.sin(gamma) -
      0.006758 * Math.cos(2 * gamma) +
      0.000907 * Math.sin(2 * gamma) -
      0.002697 * Math.cos(3 * gamma) +
      0.00148 * Math.sin(3 * gamma);

    var latRad = toRadians(FARM_LAT);

    // Hour angle at official sunrise/sunset (degrees), clamped for polar edge cases
    var cosHourAngle =
      Math.cos(toRadians(OFFICIAL_ZENITH_DEGREES)) / (Math.cos(latRad) * Math.cos(declinationRad)) -
      Math.tan(latRad) * Math.tan(declinationRad);
    cosHourAngle = Math.max(-1, Math.min(1, cosHourAngle));
    var hourAngleDeg = toDegrees(Math.acos(cosHourAngle));

    // Minutes from midnight UTC of the requested day (sunset may exceed 1440)
    var sunriseUtcMinutes =
      SOLAR_NOON_UTC_MINUTES - MINUTES_PER_LONGITUDE_DEGREE * (FARM_LNG + hourAngleDeg) - eqTimeMinutes;
    var sunsetUtcMinutes =
      SOLAR_NOON_UTC_MINUTES - MINUTES_PER_LONGITUDE_DEGREE * (FARM_LNG - hourAngleDeg) - eqTimeMinutes;

    return {
      sunriseMs: dayStartMs + sunriseUtcMinutes * MS_PER_MINUTE,
      sunsetMs: dayStartMs + sunsetUtcMinutes * MS_PER_MINUTE
    };
  }

  /**
   * True if nowMs (epoch ms) falls inside the daylight window of EITHER the
   * current UTC calendar day or the previous one. The two-day check is
   * load-bearing: a Michigan summer evening (e.g. 20:30 EDT = 00:30 UTC) has
   * already rolled to the next UTC day, whose own window hasn't started —
   * daylight there belongs to the PREVIOUS UTC day's window.
   *
   * @param {number} nowMs epoch milliseconds
   * @returns {boolean}
   */
  function isDaylight(nowMs) {
    if (!Number.isFinite(nowMs)) {
      throw new TypeError('isDaylight: nowMs must be a finite epoch-ms number');
    }

    function isInWindow(ms, dateForDay) {
      var win = getSunWindowUtc(
        dateForDay.getUTCFullYear(),
        dateForDay.getUTCMonth() + 1,
        dateForDay.getUTCDate()
      );
      return ms >= win.sunriseMs && ms < win.sunsetMs;
    }

    return (
      isInWindow(nowMs, new Date(nowMs)) ||
      isInWindow(nowMs, new Date(nowMs - MS_PER_DAY))
    );
  }

  var SunTimes = {
    FARM_LAT: FARM_LAT,
    FARM_LNG: FARM_LNG,
    getSunWindowUtc: getSunWindowUtc,
    isDaylight: isDaylight
  };

  root.SunTimes = SunTimes;
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = SunTimes;
  }
})(typeof window !== 'undefined' ? window : globalThis);
